### DISCUSSION TITLE
Phoniebox + Homepod working

### DISCUSSION THREAD

**Discussion by charlesvestal** [MiczFlor/RPi-Jukebox-RFID#1766] [Ideas]
After some convoluted trial and error, I have a homepod working with my phoniebox record player. The rough steps:

- Enable snd-aloop module to get loopback devices
- Add another output outputting to the loopback device
- Make a fifo somewhere 
- Add a fifo output that arecords that loopback _input_ to the fifo file
- install owntone / forked-daapd with a modified mpd port (i used 6610), with autostart pipes enabled, connected to a homepod
- make sure the fifo file is in your owntone shares

A bit of a convoluted mess, but it seems to work – and it seems stereo pairs work with 15.4 too, which could be nice.

Would love to see native airplay support in MPD, or for the raop2 support in pulseaudio to work but until then, this seems to work.

**Comment by pabera** [comment]
Are you referring to V 2.x or 3.x?

It would be interesting to know what your steps were to get it working.

E.g. V 3.x is using PulseAudio now and already integrates with many sound cards and Bluetooth devices. Maybe that's a good way to explore further ?

**Reply by charlesvestal** [reply]
I'm on 2.x still as pause-on-remove is important to me. My testing with pulseaudio couldn't ever get it to pair with homepods (but would work with legacy airplay 1 devices). Will poke on that some in the future.

If I recall my steps were

- add snd-aloop to /etc/modules and reboot
- make a fifo in your shared directory with `mkfifo /home/pi/RPi-Jukebox-RPi-Jukebox-RFID/shared/audiofolders/piboxpipe`
- edit /etc/mpd.conf and add an additional loopback audio device
```
audio_output {
        type            "alsa"
        name            "Loopback"
       device          "hw:1,0,4"        # optional
       mixer_type      "hardware"      # optional
       mixer_device    "default"       # optional
        mixer_control   "Headphone"             # optional
       mixer_index     "0"             # optional
}
```
- add the additional fifo output to /etc/mpd.conf
```
audio_output {
 type    "fifo"
 name    "my_fifo"
 format  "44100:16:2"
}
```

- `sudo service mpd restart to enable` the new outputs
- install owntone, via these instructions: https://forums.raspberrypi.com/viewtopic.php?t=49928
- also install pulseaudio with `sudo apt install pulseaudio`
- edit /etc/owntone.conf and change `port` to 6610 so you can have multiple mpd instances and enable `pipe_autostart` so it will watch your shared folder for the pipe.
- add your shared folder to the owntown shares, i.e.         `directories = { "/srv/music", "/home/pi/RPi-Jukebox-RFID/shared/audiofolders" }`
- `sudo service owntone start`

play something on the RPi-Jukebox-RFID interface, and you'll see owntown (webui on port 3689) pick up the pipe. in owntone, select a homepod as a speaker output.

Stop the Jukebox and owntone will stop as the fifo has no data. Playing will resume and reconnect to the speaker.


Like I said, it's _very_ convoluted, and I'd much prefer to use pulseaudio if it would work as expected. But for a proof of concept, this strung together hack seems to work. :)

**Comment by charlesvestal** [comment]
Just wanted to bump this with the steps I have working reliably:

- install bullseye lite
- install RPi-Jukebox-RFID (2.x master)
`cd; rm buster-install-*; wget https://raw.githubusercontent.com/MiczFlor/RPi-Jukebox-RFID/master/scripts/installscripts/buster-install-default.sh; chmod +x buster-install-default.sh; ./buster-install-default.sh`
- immediately remove mpd
`sudo apt remove mpd`
- install owntone
```
wget -q -O - http://www.gyfgafguf.dk/raspbian/owntone.gpg | sudo gpg --dearmor --output /usr/share/keyrings/owntone-archive-keyring.gpg
sudo wget -q -O /etc/apt/sources.list.d/owntone.list http://www.gyfgafguf.dk/raspbian/owntone-bullseye.list
sudo apt update
sudo apt install owntone
```

You now have RPi-Jukebox-RFID and owntone running simultaneously with owntone's config. Now we need to make them share paths and change the format the Jukebox scripts work so owntone can use them

- edit owntone's config
`sudo nano /etc/owntone.conf`
set library directory from srv/music to “/home/pi/RPi-Jukebox-RFID/shared/audiofolders”
disable ipv6 
set local sound to alsa
set mixer name to “Headphone”
- edit Jukebox scripts
in playout_controls.sh change these to pass with "file:/path" format
line 905      ` mpc add file:"${VALUE}"`
line 957       `mpc add file:"${VALUE}"`
line 978       `mpc add file:"${VALUE}" `

in rfid_trigger_play.sh, pass the full path 
line 500 	        `$PATHDATA/playout_controls.sh -c=playlistaddplay -v="$AUDIOFOLDERSPATH/$FOLDER"`

- reboot
- add some files to audiofolders
- in owntone (port 3689), choose a homepod or airplay target
- use Jukebox as normal. owntone's mpd will maintain the airplay target and the jukebox will pass the files as expected

hope this helps someone! 

note: i couldn't quit make this work with my apple tv and paired homepods, but unpaired homepods work great, including a lone homepod mini, and if you play to both left and right of a stereo pair (they appear individually) from owntone it appears to actually play stereo.
